### DISCUSSION TITLE
Perplexity with nltk and TorchMetrics

### DISCUSSION THREAD

**Discussion by d-kleine** [rasbt/MachineLearning-QandAI-book#9] [Q&A] [answered]
With reference to the given example in [this notebook](https://github.com/rasbt/MachineLearning-QandAI-book/blob/main/supplementary/q19-evaluation-llms/perplexity.ipynb), would it be possible to demonstrate how perplexity can be calculated with 
- [nltk](https://stackoverflow.com/questions/54941966/how-can-i-calculate-perplexity-using-nltk)

There is also perplexity implemented in [TorchMetrics](https://torchmetrics.readthedocs.io/en/stable/text/perplexity.html), but it seems to take the log probabilities and ground truth values, different to the example you have provided.

**Comment by rasbt** [comment]
Good suggestions! As far as I know, NLTK doesn't implement perplexity functions ... Torchmetrics does though. I added an example and the results are pretty close. I wonder if the discrepancy is due to rounding error.

**Reply by d-kleine** [reply]
Thank you! That `torchmetrics` example is way better for understanding from my point of view!  👍 ❤️ 

EDIT: `nltk` also seems to provide a perplexity function with its API, see [here](https://www.nltk.org/api/nltk.lm.api.html#nltk.lm.api.LanguageModel.perplexity)

One thing which I did not understand are the values for the model outputs:
- I understand that there is semantic similarity for some word pairs like "cat"-"fox" because they are both animals. Why do word pairs like "black"-"quick" have a probality score of 10% (`0.1`), some other words pairs like "fast"-"lazy" only score of 1% (`0.01`)?
- Also, some word probabilities don't add up to 100% for the vocabulary, see 
```python
[0.0, 0.0, 0.0, 0.0, 0.0, 0.1, 0.01, 0.01, 0.00, 0.85, 0.00, 0.00] # fast, index 9
```
= 0.97 - is this correct?

Those small values could be a reason why there is a small discrepancy to the score of the `numpy` implementation.

Another thing I have noticed:

```python
perp = Perplexity()
perp(torch.log(model_outputs), targets)
```

`torch.log()` takes the base of e, see [here](https://pytorch.org/docs/stable/generated/torch.log.html), but in your implementation with `numpy` you have used the logarithm with the base of 2:

```python
log_probs = np.log2(probabilities)
```

Does this needs to be changed to `torch.log2()` to use logarithm with the base of 2 as well, see [here](https://pytorch.org/docs/stable/generated/torch.log2.html)?
```python
perp = Perplexity()
perp(torch.log2(model_outputs), targets) # torch.log2() instead of torch.log()
```

This would change the score slightly as well, of course.

**Reply by rasbt** [reply]
> = 0.97 - is this correct?

Good catch here. fixing this so that it sums up to 1 fixes it so that the perplexities now match! Thanks!!

> EDIT: nltk also seems to provide a perplexity function with its API, see [here](https://www.nltk.org/api/nltk.lm.api.html#nltk.lm.api.LanguageModel.perplexity)

Thanks, again. As far as I understand, it would require fitting a specific language model though. I mean, this makes sense in a real-world application. But it would be very specific to that model then, and it would be a less general scoring function and hard to compare to the others (the manual and the torchmetrics ones). Unless I am overlooking sth and there is a way to use this in a model-independent way.

> I understand that there is semantic similarity for some word pairs like "car"-"fox" because they are both animals. Why do word pairs like "black"-"quick" have a probality score of 10% (0.1), some other words pairs like "fast"-"lazy" only score of 1% (0.01)?

Oops, sorry, I see how this is confusing. I just used arbitrary values here for simplicity. I clarified that in the notebook now.

**Reply by d-kleine** [reply]
Thank you, got it! 👍

One thing I am not happy about is that the values for "fast" add up to 108%:

```python
[0.0, 0.0, 0.0, 0.0, 0.0, 0.2, 0.02, 0.01, 0.00, 0.85, 0.00, 0.00]
```

I think it would be better to set one value to `0.15` to the one of `0.85`, like

```python
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.15, 0.0, 0.85, 0.0, 0.0]
```

or set the values more likely to a normal distribution like in reality, like

```python
[0.01, 0.02, 0.0, 0.0, 0.0, 0.02, 0.00, 0.1, 0.0, 0.85, 0.0, 0.0]
```

just to make it more clearer that probabilities need to add up to 100%.

Furthermore, what do you think about the `torch.log2()` instead of `torch.log()` question (please see above)?

**Reply by d-kleine** [reply]
@rasbt Could you please take second look into the code?

**Reply by rasbt** [reply]
Thanks for the reminder @d-kleine , I totally missed your follow-up! I went over the code examples again and fixed all the input inconsistencies (also added a check this time to ensure everything sums to 1).

The Torchmetrics example should now agree to the from scratch implementation and the PyTorch implementation (added this one in addition to the previous ones).

At the bottom, there is also a natural log and log-2 example. They should give the same results.

**Reply by d-kleine** [reply, marked as answer]
Thank you very much - great work, looks fine now! 👍 😃 

From my point of view, the `TorchMetrics` example has significantly improved the content, I now have a much better understanding how perplexity can be calculated when implementing in practice! 🏆

**Reply by rasbt** [reply]
Awesome, glad to hear! And thanks so much for the feedback suggesting these fixes and additions, I really appreciate it!
